### PR TITLE
Adding WQMC objective option

### PR DESCRIPTION
**Pull request by necludov**
Hi everyone,

@dpfau asked me to create a PR implementing our recent work [Wasserstein Quantum Monte Carlo](https://arxiv.org/abs/2307.07050). I just finished the initial implementation and tested it for Li atom. 

Results for Li atom. The left plot demonstrates the relative energy error, where the ground true energy is taken from the [psiformer paper](https://arxiv.org/abs/2211.13672). The right plot demonstrates the variance of the local energy, which indicates the convergence to the ground energy eigenstate. All the values are smoothed with the running average over 100 iterations.
image

The commands to run the test.
`ferminet --config ferminet/configs/atom_test.py --config.system.atom Li --config.batch_size 256 --config.pretrain.iterations 100 --config.optim.objective wvmc`
`ferminet --config ferminet/configs/atom_test.py --config.system.atom Li --config.batch_size 256 --config.pretrain.iterations 100 --config.optim.objective vmc`

**Nota Bene:**
- I had to change the evaluation of the potential to avoid 'nans' during backpropagation. It should be equivalent, but please review it.
- I guess I didn't completely follow your style guide, even though I was trying to stick to it where I could infer it.
- I haven't yet tested this particular implementation for bigger molecules (the ones we considered in the paper). Moreover, for some hyperparameters, I observed this implementation diverging for Li, perhaps, due to the smaller batch size compared to what we used in the paper.

Please, let me know how to proceed from here, and what to improve, test, and include.

### CODE REVIEW

File: ferminet/configs/atom_test.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+      None, field_type=int)
+  with cfg.ignore_type():
+    cfg.system.set_molecule = _adjust_nuclear_charge
+    cfg.config_module = '.atom'
+  cfg.network.network_type = 'psiformer'
+  cfg.optim.iterations = 10_000
+  cfg.optim.lr.delay = 5_000
+  cfg.optim.clip_median = True
+  cfg.debug.deterministic = True
+  cfg.optim.kfac.norm_constraint = 1e-3
```
Comment (jsspencer) [inline ferminet/configs/atom_test.py:80]: Except for these hyperparameters, this looks identical to `config/atom.py`. Is that correct? If the intention is just to have a simple test, I would prefer to have a minimalist config used for Li in the PR (and don't worry about using FieldReference etc) - the simplest one is more informative to users, preferable with the WQMC objective enabled.

File: ferminet/loss.py [hunk 2]
```diff
@@ -255,3 +256,133 @@ def total_energy_jvp(primals, tangents):  # pylint: disable=unused-variable
...
+        tangents[2].atoms,
+        tangents[2].charges
+    )
+    score_primal, score_tangent = jax.jvp(score, primals, tangents)
+    
+    score_norm = jnp.linalg.norm(score_primal, axis=-1, keepdims=True)
+    median = jnp.median(constants.all_gather(score_norm))
+    deviation = jnp.mean(jnp.abs(score_norm - median))
+    mask = score_norm < (median + 5*deviation)
+    log_q_tangent_out = (aux_data.grad_local_energy*score_tangent*mask).sum(1)
```
Comment (jsspencer) [inline ferminet/loss.py:379]: Out of interest, did you try masking based upon the clipping used in `clip_local_values`?

File: ferminet/loss.py [hunk 3]
```diff
@@ -255,3 +256,133 @@ def total_energy_jvp(primals, tangents):  # pylint: disable=unused-variable
...
+        tangents[0],
+        tangents[2].positions,
+        tangents[2].spins,
+        tangents[2].atoms,
+        tangents[2].charges
+    )
+    score_primal, score_tangent = jax.jvp(score, primals, tangents)
+    
+    score_norm = jnp.linalg.norm(score_primal, axis=-1, keepdims=True)
+    median = jnp.median(constants.all_gather(score_norm))
```
Comment (jsspencer) [inline ferminet/loss.py:376]: [aside, this is something we need to do at some point, no need for changes] - can improve things here by reducing the number of independent collective communication calls.

**Review by jsspencer — COMMENTED** [review]
Thanks for this! Really enjoyed the paper and results look impressive.

Don't worry about the code style -- any issues can be fixed up later.

File: ferminet/hamiltonian.py [hunk 4]
```diff
@@ -144,15 +144,17 @@ def grad_phase_closure(x):
...
-      between electrons i and j. Other elements in the final axes are not
-      required.
+    pos: Shape (neletrons, ndim). Electron positions.
   """
-  return jnp.sum(jnp.triu(1 / r_ee[..., 0], k=1))
+  nelectrons = pos.shape[0]
+  r_ee = jnp.sqrt(jnp.sum((pos[None, ...] - pos[:, None] + 
+      jnp.eye(nelectrons)[..., None])**2, axis=-1))
+  r_ee = r_ee[jnp.triu_indices_from(r_ee, 1)]
+  return (1./r_ee).sum()
```
[the remark this answers is not in the document]
Comment (necludov) [reply ferminet/hamiltonian.py:156]: Yes, you are right! That's the aftermath of my NaNs hunt. You cannot use `jnp.triu(r_ee, k=1)`, though. I've fixed it and have tested it.

File: ferminet/networks.py
[the remark this answers is not in the document]
Comment (necludov) [reply ferminet/networks.py:1]: Correct, I reverted and tested.

File: ferminet/configs/atom_test.py (hunk 1 above)
Comment (necludov) [reply ferminet/configs/atom_test.py:80]: Yes, the goal was just to create another config to test WQMC. I believe that FieldReference is the only unnecessary field here. Do you want me to delete this config?

File: ferminet/loss.py (hunk 2 above)
Comment (necludov) [reply ferminet/loss.py:379]: For sure, I did, and, as I remember, it was less stable than the current version. I haven't properly documented it, though. I can run some tests to check it.

**Comment by necludov** [comment]
I'm glad you enjoyed the paper! It means a lot coming from you! Let me know what else to check and fix.

File: ferminet/loss.py [hunk 5]
```diff
@@ -255,3 +256,133 @@ def total_energy_jvp(primals, tangents):  # pylint: disable=unused-variable
...
+    else:
+      diff = aux_data.local_energy - loss
+
+    def log_q(_params, _pos, _spins, _atoms, _charges):
+      out = batch_network(_params, _pos, _spins, _atoms, _charges)
+      kfac_jax.register_normal_predictive_distribution(out[:,None])
+      return out.sum()
+
+    score = jax.grad(log_q, argnums=1)
+    primals = (primals[0], data.positions, data.spins, data.atoms, data.charges)
```
Comment (jsspencer) [inline ferminet/loss.py:365]: Clearer and more robust (e.g. if we ever change the args order to `total_energy`) to pass in `params` instead of primals[0] here

File: ferminet/loss.py (hunk 2 above)
Comment (jsspencer) [reply ferminet/loss.py:379]: Interesting, can see it going both ways but would also have thought clipping based on `score_norm` would be better.

File: ferminet/hamiltonian.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jsspencer) [reply ferminet/hamiltonian.py:156]: Ah, I (think I) meant `jnp.sum(jnp.triu(1./(r_ee[..., 0] + jnp.eye(...)), k=1)` but doing `jnp.triu_indices_from` works well (and is probably faster). Either way, the key is to avoid a 1/0 entirely -- by adding the identity to the diagonal (which is then dropped by the `jnp.triu`) or, as you did, by doing the triu before the reciprocal. This is much nicer, thanks!

File: ferminet/hamiltonian.py [hunk 6]
```diff
@@ -173,25 +174,25 @@ def potential_nuclear_nuclear(charges: Array, atoms: Array) -> jnp.ndarray:
     charges: Shape (natoms). Nuclear charges of the atoms.
     atoms: Shape (natoms, ndim). Positions of the atoms.
   """
-  r_aa = jnp.linalg.norm(atoms[None, ...] - atoms[:, None], axis=-1)
+  natoms = atoms.shape[0]
+  r_aa = jnp.sqrt(jnp.sum((atoms[None, ...] - atoms[:, None] + 
+      jnp.eye(natoms)[..., None])**2, axis=-1))
```
Comment (jsspencer) [inline ferminet/hamiltonian.py:179]: Can you also undo this change so we just jnp.linalg.norm still here (just less surprising to the reader) please?

File: ferminet/hamiltonian.py [hunk 7]
```diff
@@ -173,25 +174,25 @@ def potential_nuclear_nuclear(charges: Array, atoms: Array) -> jnp.ndarray:
...
   return jnp.sum(
       jnp.triu((charges[None, ...] * charges[..., None]) / r_aa, k=1))
 
 
-def potential_energy(r_ae: Array, r_ee: Array, atoms: Array,
+def potential_energy(r_ae: Array, pos: Array, atoms: Array,
```
Comment (jsspencer) [inline ferminet/hamiltonian.py:184]: Can you undo this variable rename (and other changes in this function)? Then any calls using kwargs into the library aren't broken

File: ferminet/configs/atom_test.py (hunk 1 above)
Comment (jsspencer) [reply ferminet/configs/atom_test.py:80]: How about renaming it to `li_wqmc.config` and set

```
cfg.system.atom = 'Li'
cfg.optim.objective = 'wqmc'
```

or delete it, as you prefer.

[I now realise `FieldReference` on `spin_polarisation` is necessary - it allows to set the field to either `None` or an integer without `ml_collections.ConfigDict` complaining about type safety...]

**Comment by necludov** [comment]
I implemented the comments and tested them. The command to run now is
`ferminet --config ferminet/configs/li_wqmc.py --config.batch_size 256`

**Review by jsspencer — APPROVED** [review]
Looks great, thanks!
